NIFI-4759 Fixed a bug that left a hard-coded reference to _id in as t…

MikeThomsen opened the pull request:
…he update key for MongoDB upserts.

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

mgaido91 commented on nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/PutMongo.java:
```diff
@@ -155,12 +155,7 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
-                    StreamUtils.fillBuffer(in, content, true);
-                }
-            });
+            session.read(flowFile, in -> StreamUtils.fillBuffer(in, content, true));
```
this is causing the error on Ci because of the unused import which introduces. Since this is not related with the PR, can we get it back to its original state?

MikeThomsen commented:
Thanks @mgaido91. Unfortunately, the user got back to me and told me they noticed another bug that is related to this so I have to expand on it. I'll incorporate your feedback into that.

MikeThomsen commented:
@mgaido91 
@mattyb149 

Ok, added a UT that really kicks the tires on both problems. I updated the ticket with a new title and an explanation of both issues. Most of the code here is focused on handling ObjectIDs correctly.

https://issues.apache.org/jira/browse/NIFI-4759

mgaido91 commented on nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/test/java/org/apache/nifi/processors/mongodb/PutMongoTest.java:
```diff
@@ -256,4 +258,72 @@ public void testUpsertWithOperators() throws Exception {
+    @Test
+    public void testNiFi_4759_Regressions() {
+        String[] upserts = new String[]{
+                "{\n" +
```
I think having the 3 JSONs one for each row would improve readability

MikeThomsen replied:
Not following you here.

mgaido91 replied:
I just meant that is hard to read and understand the content of this JSON. I think that something like:
```
String[] upserts = new String[]{
    "{...}",
    "{...}",
    "{...}"};
```
would improve readability.

mgaido91 commented on nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/PutMongo.java:
```diff
@@ -173,13 +166,25 @@ public void process(final InputStream in) throws IOException {
+                    try {
+                        keyVal = new ObjectId((String) keyVal);
+                    } catch (Exception ex) {
+                        getLogger().error("{} is not a valid ObjectID, using raw value.", new Object[]{keyVal});
```
what about `keyVal + "is not a valid ObjectID, using raw value."`

MikeThomsen replied:
That's the general pattern that is used in other places for building log statements.

mgaido91 commented on nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/PutMongo.java:
```diff
@@ -173,13 +166,25 @@ public void process(final InputStream in) throws IOException {
+                query = new Document(updateKey, keyVal);
 
                 if (updateMode.equals(UPDATE_WITH_DOC.getValue())) {
+
```
nit: useless blank line

mgaido91 commented on nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/PutMongo.java:
```diff
@@ -173,13 +166,25 @@ public void process(final InputStream in) throws IOException {
+                final Document query;
+
+                Object keyVal = ((Map)doc).get(updateKey);
+                if (updateKey.equals("_id")) {
```
add ` && ObjectId.isValid(keyVal)`

mgaido91 commented on nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/main/java/org/apache/nifi/processors/mongodb/PutMongo.java:
```diff
@@ -173,13 +166,25 @@ public void process(final InputStream in) throws IOException {
                 final boolean upsert = context.getProperty(UPSERT).asBoolean();
                 final String updateKey = context.getProperty(UPDATE_QUERY_KEY).getValue();
-                final Document query = new Document(updateKey, ((Map)doc).get(updateKey));
+                final Document query;
```
this can be moved later where it is created

mgaido91 commented on nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/test/java/org/apache/nifi/processors/mongodb/PutMongoTest.java:
```diff
@@ -256,4 +258,72 @@ public void testUpsertWithOperators() throws Exception {
+     * * _ids that are arbitrary strings should be still go in as strings.
+     *
+     */
+
```
nit: useless blank line

mgaido91 commented on nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/test/java/org/apache/nifi/processors/mongodb/PutMongoTest.java:
```diff
@@ -256,4 +258,72 @@ public void testUpsertWithOperators() throws Exception {
+            runner.setProperty(PutMongo.MODE, "update");
+            runner.setProperty(PutMongo.UPSERT, "true");
+            runner.setProperty(PutMongo.UPDATE_QUERY_KEY, updateKeyProps[index]);
+            for (int x = 0; x < 5; x++) {
```
why doing this 5 times?

MikeThomsen replied:
Just a validation that if we run things multiple times the same result happens.

mgaido91 replied:
then 2 is enough, isn't it?

mgaido91 commented on nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/test/java/org/apache/nifi/processors/mongodb/PutMongoTest.java:
```diff
@@ -256,4 +258,72 @@ public void testUpsertWithOperators() throws Exception {
+        }
+    }
+
+    /*
```
this comment can be removed, it seems useless to me

MikeThomsen replied:
I prefer to leave it in to explicitly mark the ending of the regression testing.

mgaido91 commented on nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-processors/src/test/java/org/apache/nifi/processors/mongodb/PutMongoTest.java:
```diff
@@ -166,6 +167,7 @@ public void testUpdateDoesNotInsert() throws Exception {
         byte[] bytes = documentToByteArray(doc);
 
         runner.setProperty(PutMongo.MODE, "update");
+        runner.setProperty(PutMongo.UPSERT, "false");
```
why do we need to add this?

MikeThomsen replied:
We don't.

mgaido91 replied:
can we remove it then, please?

MikeThomsen commented:
@mgaido91 

I screwed up the pull request because of a mistake with a rebase. New PR is here:

https://github.com/apache/nifi/pull/2401
